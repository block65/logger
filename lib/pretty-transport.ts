import { createColors } from 'colorette';
import { once } from 'events';
import build from 'pino-abstract-transport';
import SonicBoom, { SonicBoomOpts } from 'sonic-boom';
import { isatty } from 'tty';
import util from 'util';
import { LogDescriptor, LogLevelNumbers } from './types.js';

export interface PrettyTransportOptions {
  destination: number | SonicBoomOpts['dest'];
}

export function createPrettifier(options?: { fd?: number; color?: boolean }) {
  const forceNoColor: boolean =
    process.env.TERM === 'dumb' || 'NO_COLOR' in process.env;

  const useColor = forceNoColor
    ? false
    : options?.color === true || (!!options?.fd && isatty(options.fd));

  const {
    bgRed,
    bgYellow,
    black,
    blue,
    bold,
    dim,
    gray,
    green,
    red,
    whiteBright,
  } = createColors({ useColor });

  const formatLevel = (level: LogLevelNumbers): string => {
    switch (level) {
      case LogLevelNumbers.Fatal:
        return bgRed(whiteBright(bold('FATAL')));
      case LogLevelNumbers.Error:
        return red('ERROR');
      case LogLevelNumbers.Warn:
        return bgYellow(black(bold('WARN')));
      case LogLevelNumbers.Info:
        return blue('INFO');
      case LogLevelNumbers.Debug:
        return green('DEBUG');
      case LogLevelNumbers.Trace:
        return dim('TRACE');
      case LogLevelNumbers.Silent:
      default:
        return dim('SILENT');
    }
  };

  return (log: LogDescriptor): string => {
    const { level, msg = '', time, name, hostname, pid, ...rest } = log;

    const formattedName = name ? `(${name})` : '';
    const formattedMsg = msg && ' ' + bold(msg);

    const formattedRest =
      Object.keys(rest).length > 0
        ? ' ' +
          util.inspect(rest, {
            colors: useColor,
            compact: true,
            sorted: true,
          })
        : '';

    return `${gray(new Date(time).toJSON())} ${formatLevel(level).padEnd(
      6,
      ' ',
    )}${formattedName} ${formattedMsg}${formattedRest}\n`;
  };
}

export default async function (options: PrettyTransportOptions) {
  const fd =
    typeof options.destination === 'number' ? options.destination : undefined;

  const prettifier = createPrettifier({ fd });

  // SonicBoom is necessary to avoid loops with the main thread.
  // It is the same of pino.destination().
  const destination = new SonicBoom({
    fd,
    dest:
      typeof options.destination !== 'number'
        ? options.destination?.toString()
        : undefined,
    sync: false,
  });

  await once(destination, 'ready');

  return build(
    async function (source) {
      for await (let obj of source) {
        const toDrain = !destination.write(prettifier(obj) + '\n');
        // This block will handle backpressure
        if (toDrain) {
          await once(destination, 'drain');
        }
      }
    },
    {
      async close(err?: Error) {
        if (err) {
          console.warn(err);
        }
        destination.end();

        await once(destination, 'close');
      },
    },
  );
}
