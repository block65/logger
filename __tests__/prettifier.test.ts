import { beforeEach, describe, expect, jest, test } from '@jest/globals';
import { createPrettifier } from '../lib/pretty-transport.js';

describe('Prettifier', () => {
  jest.useRealTimers();

  const oldEnv = process.env;
  beforeEach(() => {
    jest.resetModules();
    process.env = { ...oldEnv };
  });

  test('No Color', async () => {
    process.env.NO_COLOR = 'true';

    const m = await import('../lib/pretty-transport.js');

    const prettifier = m.createPrettifier();

    expect(
      prettifier({
        level: 20,
        pid: 1,
        time: 1234567890,
        hostname: 'test',
        msg: 'kek',
      }),
    ).toMatchSnapshot();

    expect(
      prettifier({
        level: 20,
        pid: 1,
        time: 1234567890,
        hostname: 'test',
        msg: 'kek',
        random: {
          bool: true,
          str: 'hello',
          number: 12,
          arr: [1, 2, 3],
        },
      }),
    ).toMatchSnapshot();
  });

  test('With colour', async () => {
    const m = await import('../lib/pretty-transport.js');

    const prettifier = m.createPrettifier({
      color: true,
    });

    expect(
      prettifier({
        level: 20,
        pid: 1,
        time: 1234567890,
        hostname: 'test',
        msg: 'kek',
      }),
    ).toMatchSnapshot();

    expect(
      prettifier({
        level: 20,
        pid: 1,
        time: 1234567890,
        hostname: 'test',
        msg: 'kek',
        random: {
          bool: true,
          str: 'hello',
          number: 12,
          arr: [1, 2, 3],
        },
      }),
    ).toMatchSnapshot();
  });

  test('No colour on non tty fd', async () => {
    const prettifier = createPrettifier({
      fd: 1337,
    });

    expect(
      prettifier({
        level: 20,
        pid: 1,
        time: 1234567890,
        hostname: 'test',
        msg: 'kek',
        random: {
          bool: true,
          str: 'hello',
          number: 12,
          arr: [1, 2, 3],
        },
      }),
    ).toMatchSnapshot();
  });
});
