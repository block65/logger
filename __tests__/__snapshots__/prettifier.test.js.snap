// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Prettifier No Color 1`] = `
"1970-01-15T06:56:07.890Z DEBUG   kek
"
`;

exports[`Prettifier No Color 2`] = `
"1970-01-15T06:56:07.890Z DEBUG   kek { random: { arr: [ 1, 2, 3 ], bool: true, number: 12, str: 'hello' } }
"
`;

exports[`Prettifier No colour on non tty fd 1`] = `
"1970-01-15T06:56:07.890Z DEBUG   kek { random: { arr: [ 1, 2, 3 ], bool: true, number: 12, str: 'hello' } }
"
`;

exports[`Prettifier With colour 1`] = `
"[90m1970-01-15T06:56:07.890Z[39m [32mDEBUG[39m  [1mkek[22m
"
`;

exports[`Prettifier With colour 2`] = `
"[90m1970-01-15T06:56:07.890Z[39m [32mDEBUG[39m  [1mkek[22m { random: { arr: [ [33m1[39m, [33m2[39m, [33m3[39m ], bool: [33mtrue[39m, number: [33m12[39m, str: [32m'hello'[39m } }
"
`;
